soc: xlnx: zynqmp: Support I/D caches

robhancocksed opened the pull request:
Support enabling the instruction and data caches in the Cortex-R5 on the Xilinx ZynqMP RPU platform when CONFIG_CACHE_MANAGEMENT is enabled.

Requested in https://github.com/zephyrproject-rtos/zephyr/issues/79990

robhancocksed committed:
arch: arm: cortex_a_r: Cortex-R5 has I/D caches

Cortex-R5 has instruction and data caches present so update Kconfig
accordingly.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed committed:
soc: xlnx: zynqmp: Enable I/D caches

On ZynqMP, the RPU Cortex-R5 cores come up by default without
instruction and data caches enabled. Enable them as part of
soc_early_init_hook when CONFIG_CACHE_MANAGEMENT is enabled.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

ibirnbaum commented:
If the caches can now be enabled for the ZynqMP's R-Cores, this should be in sync with the indication that nocache memory is supported. I'll have to consider that in #79221 and add a matching placement directive in the GEM Ethernet driver, otherwise, the Ethernet DMA will stop working once the caches are enabled.

robhancocksed commented:
> If the caches can now be enabled for the ZynqMP's R-Cores, this should be in sync with the indication that nocache memory is supported. I'll have to consider that in #79221 and add a matching placement directive in the GEM Ethernet driver, otherwise, the Ethernet DMA will stop working once the caches are enabled.

I didn't automatically enable the CONFIG_CACHE_MANAGEMENT option for any ZynqMP boards yet, partially for that reason - any drivers doing DMA would need to ensure they mark data structures as uncached or explicitly flush/invalidate them as needed, so the user probably should opt in at this point.

For the Ethernet driver, using an uncached region is probably good enough for the descriptors, but may be less than ideal for the packet buffers compared to doing explicit flush/invalidate operations on regular cached memory, as doing reads/writes to an uncached area a byte/word at a time is probably a lot slower than a whole cache line at a time (depending on how fast the memory in question is to begin with).

ibirnbaum commented:
@wearyzen Yes, if the CONFIG_CACHE_MANAGEMENT switch isn't permanently on yet, all potential issues can be worked out breaking anything immediately.

@robhancocksed LGTM, and I'm adding to my to-do list: add the ICACHE and DCACHE config items for the A9 as well, check if the arch cache management code works as expected on Cortex-A (which is likely as it only supports L1 operations which should be similar on both cores, the L2 in the Zynq can't be used anyway), and then fix up the cache initialization for the Zynq-7000 a little. I checked the TRM recently and it seems that the cache flush/enable order in conjunction with the MMU setup probably isn't entirely up to spec. So far, the L1 caches only get disabled once in soc_reset_hook() and get enabled once after the MMU setup, in both cases by directly modifying the bits of the SCTLR register. Calling well defined cache API functions is probably preferable, in addition to making sure that all that happens in the right order. All cache-related operations in the Zynq just bypass the Kconfig items completely at the time being...

Also, I'll have a look at the CONFIG_NOCACHE implementation for the Cortex-R's MPU right away. /If/ the region size bit mask calculation works as intended and a matching alignment/region size padding is specified in the linker command file, that might not be such a big deal. With the Cortex-A, this memory region is definitely considered and set up in the MMU accordingly.

robhancocksed commented:
Seems there's a test failure now, in the llext.readonly_mpu test on the qemu_cortex_r5/zynqmp_rpu target. It looks like if something selects the LLEXT subsystem, as that test does, CONFIG_CACHE_MANAGEMENT gets selected as well, so this test does run with I/D caches enabled now. I was able to reproduce that failure locally.

I was unable to attach gdb to the QEMU running on the R5 (neither the Zephyr SDK aarch64 or arm GDB variants seem to be able to connect to it - they both seem confused by QEMU reporting it's both aarch64 and armv7 and return different errors, and don't report a proper execution state). However, I was able to attach the Xilinx xsdb debugger, but wasn't able to conclusively tell what's going on. It doesn't look like it has crashed in any way, almost like it's running and making progress, but seems to be stalling occasionally until you do a "stop" and "continue" in the debugger. Or possibly it's just running extremely slowly in general? The behavior is very strange, i.e. you can run for a few seconds, stop again, and it's still at the exact same PC, but doing a few "stpi" instructions to step an instruction at a time before continuing allows it to keep making progress?

Strangely enough, this only seems to happen when both I and D caches are enabled. Disabling either one of them seems to avoid this problem and the test runs to completion.

This is a fairly special case where the llext code is loading in a loadable module, and then flushing/invalidating I/D cache lines as appropriate to ensure the CPU is seeing a valid view of the new data. Given that QEMU is involved here, I'm wondering if something is tickling a bug in the Xilinx QEMU R5 simulation when the caches are enabled.

robhancocksed committed:
arch: arm: cortex_a_r: Flush entire I-cache on arch_icache_invd_range

Previously, the arch_icache_invd_range function simply returned
-ENOTSUP for Cortex A/R. However, this causes a correctness issue when
the processor modifies memory and then executes it as instructions, like
the LLEXT subsystem does. The CPU might not see the up-to-date instructions
in memory.

This caused an LLEXT test failure when running on an emulated Cortex-R5
in QEMU with caches enabled. QEMU doesn't actually emulate the cache,
but it does seem to need the ISB barrier that is executed as part of the
cache invalidate operation in order to handle self-modifying code
properly.

These CPUs do in fact support a selective I-cache invalidate operation
(ICIMVAU) but the CMSIS library doesn't currently support it. For now,
just invalidate the entire I-cache when an icache_invd_range operation
is performed.

Signed-off-by: Robert Hancock <robert.hancock@calian.com>

robhancocksed commented:
It looks like the problem was that LLEXT used selective instruction cache flush operations using sys_cache_instr_invd_range, but that ended up being a NOP for Cortex-A/R processors. These CPUs do actually support a selective instruction cache flush using the ICIMVAU operation, but the CMSIS library doesn't currently support it. However, for proper operation, we should at least do a full instruction cache invalidate when a selective invalidate is requested. I've added another commit to make this change.

As far as why QEMU was having problems, despite not emulating the cache, it does appear to need the ISB operation that's done as part of instruction cache invalidation in order to handle "self-modifying code" situations properly. Just adding a barrier_isync_fence_full call to the arch_icache_invd_range function seems enough to satisfy QEMU (but wouldn't be enough on hardware with a real instruction cache). Still not certain why it worked in QEMU without any of these changes, however - maybe some other QEMU behavior change with cache enabled that I'm not aware of..

ibirnbaum commented:
> I was unable to attach gdb to the QEMU running on the R5 (neither the Zephyr SDK aarch64 or arm GDB variants seem to be able to connect to it - they both seem confused by QEMU reporting it's both aarch64 and armv7 and return different errors, and don't report a proper execution state).

Despite executing code on the R5-cores, QEMU will identify itself as aarch64. Using the GDB console, I somehow managed to invoke a second CPU partition which contains the R-cores which accepts aarch32 binaries. Although I could then load the image, single step or even just regular execution didn't work, I got nothing useful out of that. Even if the partition switch had worked, I don't think that it's possible to invoke it right from the command line, and I don't know enough about GDB in order to tell if there's a chance of passing settings to GDB via a config file, either via the build toolchain or by just having it in the board directory. When things get rough, I go for the Lauterbach right away...

QEMU's integrated GDB server works fine for the Zynq in qemu_cortex_a9, but the multi-arch setup of the ZynqMP seems to be less straightforward.

ibirnbaum commented:
@robhancocksed I have the nocache area setup for the ZynqMP's MPU worked out, but I'll add that separately after #79221 will hopefully be merged, as the additional mapping entry is only one contributing factor besides the aarch32 Cortex-A/R linker command file and the Kconfig settings. That's beyond the scope of what the MPU setup overhaul PR is supposed to achieve. So nocache should eventually become available even if the ICache/DCache being active becomes the default on the ZynqMP. I'll verify that the MPU setup / cache enable order is correct and then post my review.

wearyzen commented on arch/arm/core/cortex_a_r/cache.c:
```diff
@@ -208,7 +208,13 @@ int arch_icache_flush_range(void *start_addr, size_t size)
+	 * the instruction cache, but not currently supported by CMSIS.
+	 * For now, invalidate the entire cache.
+	 */
+	L1C_InvalidateICacheAll();
```
could you wrap this under CPU_CORTEX_R5?

robhancocksed commented:
Ping? Looks like this needs an additional review.

ibirnbaum reviewed (approved):
Verified via JTAG debugging that the MPU / ICache / DCache init order is correct, corresponding SCTLR enable bits are set, an average test application without nocache requirements works as intended :).
